### PR TITLE
Update and add types for released versions of FHIR.

### PR DESCRIPTION
**Pull request by bkaney**
This PR sets the current version of FHIR to current version R4. It also includes older versions, R2 and R3, as they are still used in the industry. 

This also makes it clear this is a non-npm package, instead types that follow the HL7 FHIR specification.

---

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: http://hl7.org/fhir/directory.html
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by bkaney** [comment]
Hi @yinzara  - the fhir-kit-client types are blocking this from passing CI. I am going to create official types in the library. Do you think we can pull @types/fhir-kit-client in the meantime (see: https://github.com/DefinitelyTyped/DefinitelyTyped#removing-a-package)? 

But I think I am in a catch-22 here. I will start a pr in fhir-kit-client for type support today/tomorrow so it should be quick to update. Otherwise, I am open to other suggestions...

**Comment by jmandel** [comment]
Thanks @bkaney -- really looking forward to having these FHIR types published!

**Comment by oliveregger** [comment]
thanks @bkaney, also looking forward to work with DefinitelyTyped/TypeScript and FHIR R4

**Comment by bkaney** [comment]
> Hi @yinzara - the fhir-kit-client types are blocking this from passing CI. I am going to create official types in the library. Do you think we can pull @types/fhir-kit-client in the meantime (see: https://github.com/DefinitelyTyped/DefinitelyTyped#removing-a-package)?
> 
> But I think I am in a catch-22 here. I will start a pr in fhir-kit-client for type support today/tomorrow so it should be quick to update. Otherwise, I am open to other suggestions...
> 
> (FWIW - I am not using namespaces, as that practice seems to be frowned upon for external use - https://www.typescriptlang.org/docs/handbook/namespaces-and-modules.html#pitfalls-of-namespaces-and-modules)

Update - there is a patch to fhir-kit-client types to decouple from a particular version of FHIR

File: types/fhir-kit-client/index.d.ts
Comment (jhuenges) [inline types/fhir-kit-client/index.d.ts:12]: Can this be removed?

File: types/fhir-kit-client/index.d.ts
Comment (jhuenges) [inline types/fhir-kit-client/index.d.ts:16]: Same here

**Review by jhuenges — COMMENTED** [review]
I cannot say much about the changes to `types/fhir-kit-client`, but it looks okay for me.

Regarding the other changes: I really appreciate this! Code wise it looks good to me. I tested it on the the linked [playground](https://www.typescriptlang.org/play?&dtPR=53511&install-plugin=playground-dt-review#code/JYWwDg9gTgLgBAKjgQwM5wMoBUCqBmOAMyghDgCJCALYKAegDc9yBuAWAChRJZEV0ASgBYiJMpRpRWnTgGMIAO1TxkALjjCAdAEFZ8gK4L4AXjgBvTnDhQApqgj6osm1gCeYG+oDkug0a+cAL7sHPJKEAA2NpoREADmABTIAJQhcorKcABG6tj4OnoORnCmFhxW1nYOTi7unnA+hYYwARyB6eFRMfEJWalAA) and it didn't like importing `import * as STU3 from 'fhir/v3'`. Not sure if this is a not supported by the playground or if something is not working. 
For local testing I added a `package.json` to the types directory and it worked. Maybe you have an idea?

**Comment by bkaney** [comment]
> Regarding the other changes: I really appreciate this! Code wise it looks good to me. I tested it on the the linked [playground](https://www.typescriptlang.org/play?&dtPR=53511&install-plugin=playground-dt-review#code/JYWwDg9gTgLgBAKjgQwM5wMoBUCqBmOAMyghDgCJCALYKAegDc9yBuAWAChRJZEV0ASgBYiJMpRpRWnTgGMIAO1TxkALjjCAdAEFZ8gK4L4AXjgBvTnDhQApqgj6osm1gCeYG+oDkug0a+cAL7sHPJKEAA2NpoREADmABTIAJQhcorKcABG6tj4OnoORnCmFhxW1nYOTi7unnA+hYYwARyB6eFRMfEJWalAA) and it didn't like importing `import * as STU3 from 'fhir/v3'`. Not sure if this is a not supported by the playground or if something is not working.
> For local testing I added a `package.json` to the types directory and it worked. Maybe you have an idea?

@jhuenges The way it is structured now, R2 and R3 are versions (and R4 is the current). And I don't think you can load multiple versions of type definitions at the same time, as they use the same namespace. This means to install and use R3, for instance, you would need to `npm install @types/fhir@v3` (once it is published). I don't know how that preview app works.

For applications where you need to have multiple versions of FHIR types loaded at the same time, I was planning on creating a package up a regular NPM package that exports as modules.

What do you think?

**Comment by bkaney** [comment]
Alternatively, I guess we could create multiple namespaces here in DT for each version of FHIR (e.g. `fhir2`, `fhir3`, and `fhir4`).

**Comment by jhuenges** [comment]
@bkaney thanks! I didn't know that. 

How would patching work? For example if you have installed `@types/fhir@v3` and there is a bug in the definitions.

In general I think it might be useful to somehow support multiple versions at the same time. Would be helpful when you are transitioning from STU3 to R4 for example.

**Comment by bkaney** [comment]
> In general I think it might be useful to somehow support multiple versions at the same time. Would be helpful when you are transitioning from STU3 to R4 for example.

Yes, I agree 100% with this use case. I think from a FHIR perspective this is common enough to support here in DT. I can re-structure this PR to accommodate this. Conceptually there is no type `fhir`, instead there will be `fhir4`, `fhir3`, etc. What do you think?

**Comment by jhuenges** [comment]
I am not that knowledgeable in FHIR terminology so I am not sure what a user might expect. Maybe STU3, R4, ...

**Comment by bkaney** [comment]
Okay, now there are 3 namespaces: `fhir2`, `fhir3`, and `fhir4`. The only problem now is fhir-kit-client is coupled to one (in this case `fhir4`. 

@yinzara I think we should do this now to get the PR merged .. then I will move your fhir-kit-client types into the project, and allow to make clients of various versions (and remove that from DT).

**Review by RyanCavanaugh — APPROVED** [review]
This looks OK from a DT maintainer perspective; please confirm with your co-owners. Thanks!

**Comment by bkaney** [comment]
Okay. Ready to merge.

**Pull request merged**
